File: shen-derive/verify/samples.go
// Package verify generates Go tests that check an implementation matches
// a Shen spec. The spec is the oracle: we evaluate it on sampled inputs and
// emit a test that asserts the implementation produces the same outputs.
package verify

import (
	"fmt"
	"strings"

	"github.com/pyrex41/Shen-Backpressure/shen-derive/core"
	"github.com/pyrex41/Shen-Backpressure/shen-derive/specfile"
)

// Sample is one concrete input value for a single parameter of the spec.
// It carries both the evaluator value (for spec evaluation) and the Go
// source expression (for emitting the test file).
type Sample struct {
	Value  core.Value // used by the spec evaluator
	GoExpr string     // used by the generated Go test, e.g. "mustAmount(100)"
}

// GenSamples returns a small, deterministic set of samples for the given
// Shen type. The sample count is intentionally tiny — the goal is a handful
// of meaningful cases, not exhaustive coverage.
func GenSamples(shenType string, tt *specfile.TypeTable) ([]Sample, error) {
	shenType = strings.TrimSpace(shenType)

	// list types: (list T)
	if elemType := specfile.ElemType(shenType); elemType != "" {
		elemSamples, err := GenSamples(elemType, tt)
		if err != nil {
			return nil, err
		}
		return listSamples(elemType, elemSamples, tt), nil
	}

	// primitives
	switch shenType {
	case "number":
		return numberSamples(), nil
	case "string", "symbol":
		return []Sample{
			{Value: core.StringVal(""), GoExpr: `""`},
			{Value: core.StringVal("alice"), GoExpr: `"alice"`},
			{Value: core.StringVal("bob"), GoExpr: `"bob"`},
		}, nil
	case "boolean":
		return []Sample{
			{Value: core.BoolVal(true), GoExpr: "true"},
			{Value: core.BoolVal(false), GoExpr: "false"},
		}, nil
	}

	// declared types: look up in TypeTable
	entry, ok := tt.Entries[shenType]
	if !ok {
		return nil, fmt.Errorf("unknown Shen type %q", shenType)
	}

	switch entry.Category {
	case specfile.CatWrapper, specfile.CatConstrained:
		return wrapperSamples(entry, tt)

	case specfile.CatComposite, specfile.CatGuarded:
		return compositeSamples(entry, tt)

	case specfile.CatAlias:
		// Follow the alias: shengen records WrappedType on the first premise.
		// For v1 we don't track alias targets here; treat like the underlying
		// entry if present, otherwise error.
		return nil, fmt.Errorf("alias type %q not supported in samples yet", shenType)

	case specfile.CatSumType:
		return nil, fmt.Errorf("sum type %q sampling not supported yet", shenType)

	default:
		return nil, fmt.Errorf("unhandled category %q for %q", entry.Category, shenType)
	}
}

// wrapperSamples samples a wrapper/constrained type by sampling its
// underlying primitive and filtering out any samples that would violate
// the constrained type's verified predicates. The evaluator value is just
// the primitive (the wrapping is invisible at eval time). The Go
// expression uses the shengen-generated constructor via a "mustXxx" helper.
func wrapperSamples(entry *specfile.TypeEntry, tt *specfile.TypeTable) ([]Sample, error) {
	primSamples, err := GenSamples(entry.ShenPrim, tt)
	if err != nil {
		return nil, err
	}

	// For constrained types, drop any sample whose raw primitive value
	// fails the verified predicates. This lets us include "unsafe" values
	// (negatives, zero) in the primitive pool without them causing panics
	// when wrapped.
	if entry.Category == specfile.CatConstrained {
		primSamples = filterByConstraints(entry, primSamples)
	}

	var out []Sample
	helper := "must" + entry.GoName
	for _, ps := range primSamples {
		out = append(out, Sample{
			Value:  ps.Value, // primitive values pass through unchanged
			GoExpr: fmt.Sprintf("%s(%s)", helper, ps.GoExpr),
		})
	}
	return out, nil
}

// numberSamples returns the canonical primitive-number sample pool.
// Includes zero, positive, negative, and a fractional value so specs that
// branch on any of these are exercised. Constrained wrappers filter this
// pool via their verified predicates.
func numberSamples() []Sample {
	return []Sample{
		{Value: core.IntVal(0), GoExpr: "0"},
		{Value: core.IntVal(1), GoExpr: "1"},
		{Value: core.IntVal(-1), GoExpr: "-1"},
		{Value: core.IntVal(5), GoExpr: "5"},
		{Value: core.FloatVal(2.5), GoExpr: "2.5"},
		{Value: core.IntVal(100), GoExpr: "100"},
	}
}

// filterByConstraints removes samples whose value fails any verified
// predicate on the entry. A sample passes iff every predicate evaluates
// to true when the entry's variable is bound to the sample value.
// Parse/eval errors are treated as failures (sample dropped).
func filterByConstraints(entry *specfile.TypeEntry, candidates []Sample) []Sample {
	if len(entry.Verified) == 0 || entry.VarName == "" {
		return candidates
	}
	preds := make([]core.Sexpr, 0, len(entry.Verified))
	for _, raw := range entry.Verified {
		p, err := core.ParseSexpr(raw)
		if err != nil {
			// Unparseable predicate — be safe: drop everything.
			return nil
		}
		preds = append(preds, p)
	}

	var out []Sample
	for _, s := range candidates {
		ok := true
		env := core.EmptyEnv().Extend(entry.VarName, s.Value)
		for _, p := range preds {
			result, err := core.Eval(env, p)
			if err != nil {
				ok = false
				break
			}
			b, isBool := result.(core.BoolVal)
			if !isBool || !bool(b) {
				ok = false
				break
			}
		}
		if ok {
			out = append(out, s)
		}
	}
	return out
}

// compositeSamples builds a small set of samples by picking the first sample
// of each field and producing one combined sample, then adding a couple of
// variations. This keeps cases bounded.
func compositeSamples(entry *specfile.TypeEntry, tt *specfile.TypeTable) ([]Sample, error) {
	if len(entry.Fields) == 0 {
		return nil, fmt.Errorf("composite %q has no fields", entry.ShenName)
	}

	// Sample each field.
	fieldSamples := make([][]Sample, len(entry.Fields))
	for i, f := range entry.Fields {
		fs, err := GenSamples(f.ShenType, tt)
		if err != nil {
			return nil, fmt.Errorf("field %s: %w", f.ShenName, err)
		}
		if len(fs) == 0 {
			return nil, fmt.Errorf("field %s: no samples", f.ShenName)
		}
		fieldSamples[i] = fs
	}

	// Produce one variation per unique index up to the longest field
	// sample list, so that if any field has a "tricky" sample
	// (e.g., a fractional number) it appears in at least one composite.
	maxVariations := 0
	for _, fs := range fieldSamples {
		if len(fs) > maxVariations {
			maxVariations = len(fs)
		}
	}

	helper := "must" + entry.GoName
	var out []Sample
	for v := 0; v < maxVariations; v++ {
		values := make([]core.Value, len(entry.Fields))
		goExprs := make([]string, len(entry.Fields))
		for i, fs := range fieldSamples {
			idx := v % len(fs)
			values[i] = fs[idx].Value
			goExprs[i] = fs[idx].GoExpr
		}
		out = append(out, Sample{
			Value:  core.ListVal(values),
			GoExpr: fmt.Sprintf("%s(%s)", helper, strings.Join(goExprs, ", ")),
		})
	}
	return out, nil
}

// listSamples returns list samples built from elem samples. It produces:
//   - the empty list
//   - a singleton for each distinct elem sample (capped)
//   - a multi-element list mixing the first few elem samples
//
// Producing one singleton per elem sample is important: otherwise
// "tricky" elem samples that only appear at higher indices (e.g. a
// fractional composite at index 3) never end up inside any list,
// leaving a whole class of bugs undetected.
func listSamples(elemType string, elemSamples []Sample, tt *specfile.TypeTable) []Sample {
	goElemType := tt.GoType(elemType)
	empty := Sample{
		Value:  core.ListVal(nil),
		GoExpr: fmt.Sprintf("[]%s{}", goElemType),
	}
	out := []Sample{empty}

	if len(elemSamples) == 0 {
		return out
	}

	const maxSingletons = 6
	n := len(elemSamples)
	if n > maxSingletons {
		n = maxSingletons
	}
	for i := 0; i < n; i++ {
		out = append(out, Sample{
			Value:  core.ListVal([]core.Value{elemSamples[i].Value}),
			GoExpr: fmt.Sprintf("[]%s{%s}", goElemType, elemSamples[i].GoExpr),
		})
	}

	// Multi-element list: first three elem samples if available.
	if len(elemSamples) >= 2 {
		m := 3
		if len(elemSamples) < m {
			m = len(elemSamples)
		}
		vals := make([]core.Value, m)
		exprs := make([]string, m)
		for i := 0; i < m; i++ {
			vals[i] = elemSamples[i].Value
			exprs[i] = elemSamples[i].GoExpr
		}
		out = append(out, Sample{
			Value:  core.ListVal(vals),
			GoExpr: fmt.Sprintf("[]%s{%s}", goElemType, strings.Join(exprs, ", ")),
		})
	}

	return out
}

File: cmd/shen-derive-ts/verify/samples.ts
// Package verify — sample generator for the TS verification harness.
//
// Ported from shen-derive/verify/samples.go. Produces deterministic
// boundary-value samples for Shen types, optionally extended with
// random draws from a seeded PRNG. Each sample carries both a runtime
// `Value` for the spec evaluator and a TS source expression for the
// emitted test file.

import {
  Env,
  boolVal,
  evalExpr,
  floatVal,
  intVal,
  listVal,
  stringVal,
  type Value,
} from "../core/eval.ts";
import { parseSexpr } from "../core/sexpr-parse.ts";
import {
  elemType,
  tsType,
  type TypeEntry,
  type TypeTable,
} from "../specfile/typetable.ts";

// --- Seeded PRNG (Mulberry32) ---
//
// We do NOT aim to match Go's math/rand byte-for-byte. The only goal
// is reproducibility within the TS harness for `--seed N`.

export class SeededRng {
  private state: number;

  constructor(seed: number) {
    // Coerce to unsigned 32-bit.
    this.state = seed >>> 0;
  }

  /** Returns a float in [0, 1). */
  nextFloat(): number {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let t = this.state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  /** Returns an integer in [min, max] inclusive. */
  nextInt(min: number, max: number): number {
    if (max < min) throw new Error(`nextInt: max (${max}) < min (${min})`);
    const span = max - min + 1;
    return min + Math.floor(this.nextFloat() * span);
  }
}

// --- Public types ---

export type Sample = {
  /** Runtime value used by the spec evaluator. */
  value: Value;
  /** TS source expression used in the generated test file. */
  tsExpr: string;
};

export type SampleCtx = {
  tt: TypeTable;
  /** null → deterministic, boundary values only. */
  rand: SeededRng | null;
  /** Number of extra random draws for primitive number/string types. */
  randomDraws: number;
};

// --- Boundary pools. MUST match Go exactly. ---
//
// Number pool mixes ints and floats — 2.5 is the load-bearing
// fractional value that catches int-vs-float truncation bugs. The
// int/float tagging survives through to the runtime `Value`, so
// constrained-wrapper filters see the correct `kind`.

const NUMBER_RAW: ReadonlyArray<
  { kind: "int" | "float"; val: number; expr: string }
> = [
  { kind: "int", val: 0, expr: "0" },
  { kind: "int", val: 1, expr: "1" },
  { kind: "int", val: -1, expr: "-1" },
  { kind: "int", val: 5, expr: "5" },
  { kind: "float", val: 2.5, expr: "2.5" },
  { kind: "int", val: 100, expr: "100" },
];

const STRING_POOL: ReadonlyArray<string> = ["", "alice", "bob"];
const BOOL_POOL: ReadonlyArray<boolean> = [true, false];

// --- Entry point ---

export function genSamples(ctx: SampleCtx, shenType: string): Sample[] {
  const t = shenType.trim();

  // list types: (list T)
  const elem = elemType(t);
  if (elem !== "") {
    const elemSamples = genSamples(ctx, elem);
    return listSamples(ctx.tt, elem, elemSamples);
  }

  // primitives
  switch (t) {
    case "number": {
      const out = numberSamples();
      if (ctx.rand !== null && ctx.randomDraws > 0) {
        out.push(...randomNumberSamples(ctx.rand, ctx.randomDraws));
      }
      return out;
    }
    case "string":
    case "symbol": {
      const out: Sample[] = STRING_POOL.map((s) => ({
        value: stringVal(s),
        tsExpr: JSON.stringify(s),
      }));
      if (ctx.rand !== null && ctx.randomDraws > 0) {
        out.push(...randomStringSamples(ctx.rand, ctx.randomDraws));
      }
      return out;
    }
    case "boolean":
      return BOOL_POOL.map((b) => ({
        value: boolVal(b),
        tsExpr: b ? "true" : "false",
      }));
  }

  // declared type: look up in the TypeTable.
  const entry = ctx.tt.get(t);
  if (entry === undefined) {
    throw new Error(`unknown Shen type "${t}"`);
  }

  switch (entry.category) {
    case "wrapper":
    case "constrained":
      return wrapperSamples(ctx, entry);
    case "composite":
    case "guarded":
      return compositeSamples(ctx, entry);
    case "alias":
      throw new Error(`alias type "${t}" not supported in samples yet`);
    case "sumtype":
      throw new Error(`sum type "${t}" sampling not supported yet`);
    default: {
      // Exhaustive check.
      const _never: never = entry.category;
      throw new Error(`unhandled category for "${t}": ${String(_never)}`);
    }
  }
}

// --- Primitive number pool ---

function numberSamples(): Sample[] {
  return NUMBER_RAW.map((n) => ({
    value: n.kind === "int" ? intVal(n.val) : floatVal(n.val),
    tsExpr: n.expr,
  }));
}

// --- Wrapper / constrained types ---
//
// Sample the underlying primitive, drop any candidate that fails the
// entry's verified predicates, and wrap the tsExpr in a
// `shenguard.mustXxx(raw)` helper call. The evaluator value passes
// through unchanged — wrapping is invisible at eval time.

function wrapperSamples(ctx: SampleCtx, entry: TypeEntry): Sample[] {
  let primSamples = genSamples(ctx, entry.shenPrim);
  if (entry.category === "constrained") {
    primSamples = filterByConstraints(entry, primSamples);
  }
  const helper = `${entry.importAlias || "shenguard"}.must${entry.tsName}`;
  return primSamples.map((ps) => ({
    value: ps.value,
    tsExpr: `${helper}(${ps.tsExpr})`,
  }));
}

function filterByConstraints(
  entry: TypeEntry,
  candidates: Sample[],
): Sample[] {
  if (entry.verified.length === 0 || entry.varName === "") {
    return candidates;
  }
  // Parse the verified predicates up front.
  const preds = [];
  for (const raw of entry.verified) {
    try {
      preds.push(parseSexpr(raw));
    } catch {
      // Unparseable predicate — be safe: drop everything.
      return [];
    }
  }

  const out: Sample[] = [];
  for (const s of candidates) {
    const env = Env.empty().extend(entry.varName, s.value);
    let ok = true;
    for (const p of preds) {
      try {
        const result = evalExpr(env, p);
        if (result.kind !== "bool" || !result.val) {
          ok = false;
          break;
        }
      } catch {
        ok = false;
        break;
      }
    }
    if (ok) out.push(s);
  }
  return out;
}

// --- Composite / guarded types ---
//
// One variation per field-sample index up to the longest field sample
// list, cycling shorter fields. Ensures "tricky" samples at higher
// indices (e.g. a fractional number) always land in at least one
// composite. Matches the Go implementation.

function compositeSamples(ctx: SampleCtx, entry: TypeEntry): Sample[] {
  if (entry.fields.length === 0) {
    throw new Error(`composite "${entry.shenName}" has no fields`);
  }

  const fieldSamples: Sample[][] = entry.fields.map((f) => {
    const fs = genSamples(ctx, f.typeName);
    if (fs.length === 0) {
      throw new Error(`field ${f.shenName}: no samples`);
    }
    return fs;
  });

  let maxVariations = 0;
  for (const fs of fieldSamples) {
    if (fs.length > maxVariations) maxVariations = fs.length;
  }

  const helper = `${entry.importAlias || "shenguard"}.must${entry.tsName}`;
  const out: Sample[] = [];
  for (let v = 0; v < maxVariations; v++) {
    const values: Value[] = new Array(entry.fields.length);
    const exprs: string[] = new Array(entry.fields.length);
    for (let i = 0; i < fieldSamples.length; i++) {
      const fs = fieldSamples[i]!;
      const idx = v % fs.length;
      values[i] = fs[idx]!.value;
      exprs[i] = fs[idx]!.tsExpr;
    }
    out.push({
      value: listVal(...values),
      tsExpr: `${helper}(${exprs.join(", ")})`,
    });
  }
  return out;
}

// --- List types ---
//
// Empty list + one singleton per elem sample (capped at 6) + one
// 3-element mix. "One singleton per elem sample" is the critical
// invariant that keeps tricky elem samples (e.g. a fractional number
// at index 4) from being silently excluded from every list.

function listSamples(
  tt: TypeTable,
  elemShenType: string,
  elemSamples: Sample[],
): Sample[] {
  const tsElemType = tsType(tt, elemShenType);
  const out: Sample[] = [
    { value: listVal(), tsExpr: `[] as ${tsElemType}[]` },
  ];

  if (elemSamples.length === 0) return out;

  const maxSingletons = 6;
  const n = Math.min(elemSamples.length, maxSingletons);
  for (let i = 0; i < n; i++) {
    const e = elemSamples[i]!;
    out.push({
      value: listVal(e.value),
      tsExpr: `[${e.tsExpr}]`,
    });
  }

  if (elemSamples.length >= 2) {
    const m = Math.min(elemSamples.length, 3);
    const vals: Value[] = [];
    const exprs: string[] = [];
    for (let i = 0; i < m; i++) {
      vals.push(elemSamples[i]!.value);
      exprs.push(elemSamples[i]!.tsExpr);
    }
    out.push({
      value: listVal(...vals),
      tsExpr: `[${exprs.join(", ")}]`,
    });
  }

  return out;
}

// --- Random draws (only invoked when ctx.rand !== null) ---
//
// Half ints, half 2-decimal-place floats in [-1000, 1000]. The
// fractional draws deliberately catch int-vs-float truncation bugs.

export function randomNumberSamples(rnd: SeededRng, n: number): Sample[] {
  const out: Sample[] = [];
  for (let i = 0; i < n; i++) {
    if (i % 2 === 0) {
      const v = rnd.nextInt(-1000, 1000);
      out.push({ value: intVal(v), tsExpr: v.toString() });
    } else {
      const cents = rnd.nextInt(-100000, 100000);
      const v = cents / 100;
      out.push({ value: floatVal(v), tsExpr: formatFloat(v) });
    }
  }
  return out;
}

export function randomStringSamples(rnd: SeededRng, n: number): Sample[] {
  const alphabet = "abcdefghijklmnopqrstuvwxyz0123456789";
  const out: Sample[] = [];
  for (let i = 0; i < n; i++) {
    const length = 1 + rnd.nextInt(0, 7); // 1..8 inclusive
    let s = "";
    for (let j = 0; j < length; j++) {
      s += alphabet[rnd.nextInt(0, alphabet.length - 1)]!;
    }
    out.push({ value: stringVal(s), tsExpr: JSON.stringify(s) });
  }
  return out;
}

// formatFloat prints a float with no trailing zeros, matching Go's
// strconv.FormatFloat(v, 'f', -1, 64) closely enough for the literals
// we emit. Integer-valued floats still render as e.g. "3" — the
// evaluator sees the float kind via `value`, not via `tsExpr`.
function formatFloat(v: number): string {
  if (Number.isInteger(v)) return v.toString();
  return v.toString();
}
